HeterogeneousCore/SonicTriton/src/TritonClient.cc

```diff
@@ -0,0 +1,318 @@
+namespace nic = ni::client;
+
+//based on https://github.com/NVIDIA/triton-inference-server/blob/v1.12.0/src/clients/c++/examples/simple_callback_client.cc
+
+template <typename Client>
+TritonClient<Client>::TritonClient(const edm::ParameterSet& params)
+    : Client(),
+      url_(params.getParameter<std::string>("address") + ":" + std::to_string(params.getParameter<unsigned>("port"))),
+      timeout_(params.getParameter<unsigned>("timeout")),
+      modelName_(params.getParameter<std::string>("modelName")),
+      modelVersion_(params.getParameter<int>("modelVersion")),
+      batchSize_(params.getParameter<unsigned>("batchSize")),
+      nInput_(params.getParameter<unsigned>("nInput")),
+      nOutput_(params.getParameter<unsigned>("nOutput")),
+      verbose_(params.getParameter<bool>("verbose")),
+      allowedTries_(params.getParameter<unsigned>("allowedTries")) {
+  this->clientName_ = "TritonClient";
+}
+
+template <typename Client>
+bool TritonClient<Client>::wrap(const nic::Error& err, const std::string& msg) const {
+  if (!err.IsOk()) {
+    //emit warning
+    edm::LogWarning("TritonServerWarning") << msg << ": " << err;
+  }
+  return err.IsOk();
+}
+
+template <typename Client>
+bool TritonClient<Client>::setup() {
+  bool status = true;
+
+  status = wrap(nic::InferGrpcContext::Create(&context_, url_, modelName_, modelVersion_, false),
+                "setup(): unable to create inference context");
+  if (!status)
+    return status;
+
+  //only used for monitoring
+  bool has_server = false;
+  if (verbose_) {
+    bool has_server = wrap(nic::ServerStatusGrpcContext::Create(&serverCtx_, url_, false),
+                           "setup(): unable to create server context");
+  }
+  if (!has_server)
+    serverCtx_ = nullptr;
+
+  std::unique_ptr<nic::InferContext::Options> options;
+  status = wrap(nic::InferContext::Options::Create(&options), "setup(): unable to create inference context options");
+  if (!status)
+    return status;
+
+  options->SetBatchSize(batchSize_);
+  for (const auto& output : context_->Outputs()) {
+    status = wrap(options->AddRawResult(output), "setup(): unable to add raw result");
+    if (!status)
+      return status;
+  }
+  status = wrap(context_->SetRunOptions(*options), "setup(): unable to set run options");
+  if (!status)
+    return status;
+
+  const auto& nicInputs = context_->Inputs();
+  nicInput_ = nicInputs[0];
+  nicInput_->Reset();
+
+  auto t1 = std::chrono::high_resolution_clock::now();
+  std::vector<int64_t> input_shape;
+  for (unsigned i0 = 0; i0 < batchSize_; i0++) {
+    float* arr = &(this->input_.data()[i0 * nInput_]);
+    status = wrap(nicInput_->SetRaw(reinterpret_cast<const uint8_t*>(arr), nInput_ * sizeof(float)),
+                  "setup(): unable to set data for batch entry " + std::to_string(i0));
+    if (!status)
+      return status;
+  }
+  auto t2 = std::chrono::high_resolution_clock::now();
+  if (!this->debugName_.empty())

```

Comment:
Here and below: why the need to explicitly write `this->` ?

Reply:
`TritonClient` is a class template that inherits from its template argument: https://github.com/cms-sw/cmssw/blob/ca83125faa8cbbb3145d6e000dd3e0adf126c007/HeterogeneousCore/SonicTriton/interface/TritonClient.h#L19-L20

The compiler doesn't check the base class for nondependent fields (https://isocpp.org/wiki/faq/templates#nondependent-name-lookup-members). Using `this->` makes it dependent, so the compiler can find inherited members and methods.

I did find a few cases where `this->` is not needed, and I'll remove those.

(There's a separate case in `TritonClientAsync::evaluate()` where `this` is passed to a lambda, and therefore `this->` must be used inside the lambda.)

Reply:
I see, thanks for the explanation.

(I think if you explicitly capture `this` by value in a lambda, you can reference methods and data members directly, without `this->`, unless they are shadowed by some other variable.)

Reply:
Thanks, I've removed all the `this->`s in the lambda.